hydrolib/core/io/net/models.py

```diff
@@ -0,0 +1,916 @@
+            reader.read_link1d2d(network._link1d2d)
+
+        ds.close()
+
+        return network
+
+    def to_file(self, file: Path) -> None:
+        """Write network to file
+
+        Args:
+            file (Path): File where _net.nc is written to.
+        """
+        writer = UgridWriter()
+        writer.write(self, file)
+
+    def link1d2d_from_1d_to_2d(
+        self, branchids: List[str] = None, polygon: GeometryList = None
+    ) -> None:
+        self._mesh1d._set_mesh1d()
+        self._mesh2d._set_mesh2d()
+
+        node_mask = self._mesh1d.get_node_mask(branchids)
+        if polygon is None:
+            polygon = self.meshkernel.mesh2d_get_mesh_boundaries_as_polygons()
+
+        self._link1d2d._link_from_1d_to_2d(node_mask, polygon=polygon)
+
+    def mesh2d_create_rectilinear_within_bounds(
+        self, extent: tuple, dx: float, dy: float
+    ) -> None:
+        self._mesh2d.create_rectilinear(extent=extent, dx=dx, dy=dy)
+
+    def mesh2d_create_triangular_within_polygon(self, polygon: mk.GeometryList) -> None:
+        raise NotImplementedError()
+
+    def mesh2d_clip_mesh(
+        self, polygon: mk.GeometryList, deletemeshoption: int = 1
+    ) -> None:
+        self._mesh2d.clip(polygon=polygon, deletemeshoption=deletemeshoption)
+
+    def mesh2d_refine_mesh(self, polygon: mk.GeometryList, level: int = 1) -> None:
+        self._mesh2d.refine(polygon=polygon, level=level)
+
+    def mesh1d_add_branch(
+        self,
+        branch: Branch,
+        name: str = None,
+        branch_order: int = -1,
+        long_name: str = None,
+    ) -> None:
+        self._mesh1d._add_branch(
+            branch=branch, name=name, branch_order=branch_order, long_name=long_name
+        )
+
+    def plot(
+        self,
+        ax: matplotlib.axes._subplots.AxesSubplot,
+        mesh1d_kwargs: dict = None,
+        mesh2d_kwargs: dict = None,
+        links1d2d_kwargs: dict = None,
+    ) -> None:
+
+        if mesh1d_kwargs is None:
+            mesh1d_kwargs = {"color": "C3", "lw": 1.0}
+        if mesh2d_kwargs is None:
+            mesh2d_kwargs = {"color": "C0", "lw": 0.5}
+        if links1d2d_kwargs is None:
+            links1d2d_kwargs = {"color": "k", "lw": 1.0}
+
+        # Mesh 1d
+        if not self._mesh1d.empty():
+            nodes1d = np.stack(
+                [self._mesh1d.mesh1d_node_x, self._mesh1d.mesh1d_node_y], axis=1
+            )
+            edge_nodes = self._mesh1d.mesh1d_edge_nodes
+            lc_mesh1d = LineCollection(nodes1d[edge_nodes], **mesh1d_kwargs)
+            ax.add_collection(lc_mesh1d)
+
+        # Mesh 2d
+        if not self._mesh2d.empty():
+            nodes2d = np.stack(
+                [self._mesh2d.mesh2d_node_x, self._mesh2d.mesh2d_node_y], axis=1
+            )
+            edge_nodes = self._mesh2d.mesh2d_edge_nodes
+            lc_mesh2d = LineCollection(nodes2d[edge_nodes], **mesh2d_kwargs)
+            ax.add_collection(lc_mesh2d)
```

Comment:
I think you could abstract this logic into a separate private function if we would unify the mesh1d_node_x and mesh2d_node_x to a single node_x (same_for node_y):

```python
class MeshBased(Protocol):
    node_x
    node_y
    edge_nodes

    def empty(self):
        """empty check."""

def add_mesh_to_plot(mesh: MeshBased, ax, styling_kwargs):
    if mesh.empty():
        return

    nodes = np.stack([mesh.node_x, mesh.node_y], axis=1)
    lc_mesh2d = LineCollection(nodes[mesh.edge_nodes], **styling_kwargs)
    ax.add_collection(lc_mesh2d)
```

then the internal code could just be 
```python
def plot(
        self,
        ax: matplotlib.axes._subplots.AxesSubplot,
        mesh1d_kwargs: dict = None,
        mesh2d_kwargs: dict = None,
        links1d2d_kwargs: dict = None,
    ) -> None:

        if mesh1d_kwargs is None:
            mesh1d_kwargs = {"color": "C3", "lw": 1.0}
        if mesh2d_kwargs is None:
            mesh2d_kwargs = {"color": "C0", "lw": 0.5}
        if links1d2d_kwargs is None:
            links1d2d_kwargs = {"color": "k", "lw": 1.0}
        
        add_mesh_to_plot(self._mesh1d, ax, mesh1d_kwargs)
        add_mesh_to_plot(self._mesh2d, ax, mesh2d_kwargs)

        # Links
        if not self._link1d2d.empty():
            faces2d = np.stack(
                [self._mesh2d.mesh2d_face_x, self._mesh2d.mesh2d_face_y], axis=1
            )
            link_coords = np.stack(
                [
                    nodes1d[self._link1d2d.link1d2d[:, 0]],
                    faces2d[self._link1d2d.link1d2d[:, 1]],
                ],
                axis=1,
            )
            lc_link1d2d = LineCollection(link_coords, **links1d2d_kwargs)
            ax.add_collection(lc_link1d2
```

For consistency sake you could move the links to an `add_links_to_plot` so all function operate at the same abstraction.

Reply:
Agreed. In time perhaps a seperate plotting for 1d and 2d could be desired, but we can always change it then.

Reply:
See for the alternative to LineCollection the proposed function a few comments above.